devmand/gateway/src/devmand/channels/cli/SshSessionAsync.cpp

```diff
@@ -78,57 +81,51 @@ void SshSessionAsync::setEvent(event* event) {
 void readCallback(evutil_socket_t fd, short what, void* ptr) {
   (void)fd;
   (void)what;
-  ((SshSessionAsync*)ptr)->readToBuffer();
+  ((SshSessionAsync*)ptr)->readSshDataToBuffer();
+  ((SshSessionAsync*)ptr)->processDataInBuffer();
 }
 
 socket_t SshSessionAsync::getSshFd() {
   return this->session.getSshFd();
 }
 
-void SshSessionAsync::readToBuffer() {
-  {
-    std::lock_guard<mutex> guard(mutex1);
-    ErrorHandler::executeWithCatch([this]() {
-      const string& output = this->session.read();
-      if (!output.empty()) {
-        readQueue.push(output);
-      }
-    });
-  }
-  condition.notify_one();
+void SshSessionAsync::readSshDataToBuffer() {
+  ErrorHandler::executeWithCatch([this]() {
+    const string& output = this->session.read();
+    if (!output.empty()) {
+      readQueue.push(output);
+    }
+  });
 }
 
 using namespace std::chrono_literals;
 
-string SshSessionAsync::readUntilOutputBlocking(string lastOutput) {
+void SshSessionAsync::checkExpectedOutput() {
+  if (not this->session.isOpen()) {
+    throw std::runtime_error("Session is closed");
+  }
   reading.store(true);
-  string result;
-  while (this->session.isOpen()) {
-    unique_lock<mutex> lck(mutex1);
-    bool satisfied = condition.wait_for(
-        lck, 1000ms, [this] { return this->readQueue.read_available() != 0; });
-
-    if (!satisfied) {
-      continue;
-    }
-
+  while (this->readQueue.read_available() != 0) {
     string output;
     if (!readQueue.pop(output) ||
-        output.empty()) { // sometimes we get the string "" or " " back, we can
-                          // ignore that ...
+        output.empty()) { // sometimes we get the string "" back, we can ignore
```

Comment:
when can that happen ?

Reply:
I can't explain when, but I saw it a few times in the debugger.  I changed it now for readability purposes, so this is fixed